### PR TITLE
feat(metrics): add media type label to object store latency metric

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As title.
- Add `store_media_type()` to identify the media type of different object store
- Add media_type label to the `operation_latency` metric

image

## Checklist

~~- [ ] I have written necessary rustdoc comments~~
~~- [ ] I have added necessary unit tests and integration tests~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: grafana/risingwave-dashboard.py [hunk 1]
```diff
@@ -99,6 +99,11 @@ def timeseries_bytes_per_sec(self, title, targets):
         return TimeSeries(title=title, targets=targets, gridPos=gridPos, unit="Bps", fillOpacity=10,
                           legendDisplayMode="table", legendPlacement="right", legendCalcs=["max"])
 
+    def timeseries_requests_per_sec(self, title, targets):
```
Comment (Li0k) [inline grafana/risingwave-dashboard.py:102]: It looks like `timeseries_ops`, can we use `timeseries_ops` directly?

File: src/object_store/src/object/mod.rs [hunk 2]
```diff
@@ -140,6 +140,15 @@ pub trait ObjectStore: Send + Sync {
     }
 
     async fn list(&self, prefix: &str) -> ObjectResult<Vec<ObjectMetadata>>;
+
+    fn store_media_type(&self) -> StoreMediaTypeE;
+}
+
+pub enum StoreMediaTypeE {
```
Comment (Li0k) [inline src/object_store/src/object/mod.rs:147]: IMHO, `StoreMediaTypeE` is a emun, but no need to assign `E`, can we use `StoreMediaType` as a normal `Type` announce, its more readable for me

File: src/object_store/src/object/mod.rs [hunk 3]
```diff
@@ -335,7 +352,9 @@ impl<OS: ObjectStore> MonitoredObjectStore<OS> {
             .operation_latency
             .with_label_values(&["read"])
             .start_timer();
-
+        if let StoreMediaTypeE::S3 = self.inner.store_media_type() {
```
Comment (Li0k) [inline src/object_store/src/object/mod.rs:355]: Shall we can use a function like `ReportWrite` or `ReportRead` to shadow the type match logic, and use it for other `StoreMediaType` uniformly ?

**Comment by hzxa21** [comment]
How is this metric different from the existing `Operation Rate` metric?

File: grafana/risingwave-dashboard.py (hunk 1 above)
Comment (StrikeW) [reply grafana/risingwave-dashboard.py:102]: This metric is intended for the high-level read and write, not for each specific operation (e.g. put, get, head, etc.)

File: src/object_store/src/object/mod.rs (hunk 3 above)
Comment (StrikeW) [reply src/object_store/src/object/mod.rs:355]: Good point.

**Comment by StrikeW** [comment]
> How is this metric different from the existing `Operation Rate` metric?

`Operation Rate` seems for all kinds of media types, it counts local disk and S3 in total. But this metric is for S3 only and is intended for the high-level read and write, not for specific S3 operations (e.g. put, get, head, etc.)

**Comment by hzxa21** [comment]
> > How is this metric different from the existing `Operation Rate` metric?
> 
> `Operation Rate` seems for all kinds of media types, it counts local disk and S3 in total. But this metric is for S3 only and is intended for the high-level read and write, not for specific S3 operations (e.g. put, get, head, etc.)

I don't get it. The added metrics you described are duplicates of the existing object store histogram metrics.
- If you want a separation of different media types, you can add a media type label in the existing metrics (`operation_latency`, `operation_size`) instead of creating new metrics of each object store type.
- If you want high-level read/write metrics instead of per operation metrics, you can just do a filter in the grafana dashboard. Example: https://github.com/singularity-data/risingwave/blob/796145df56d0656099609e3df88f5b3d57cd83e8/grafana/risingwave-dashboard.py#L347

**Comment by StrikeW** [comment]
> > > How is this metric different from the existing `Operation Rate` metric?
> > 
> > 
> > `Operation Rate` seems for all kinds of media types, it counts local disk and S3 in total. But this metric is for S3 only and is intended for the high-level read and write, not for specific S3 operations (e.g. put, get, head, etc.)
> 
> I don't get it. The added metrics you described are duplicates of the existing object store histogram metrics.
> 
> * If you want a separation of different media types, you can add a media type label in the existing metrics (`operation_latency`, `operation_size`) instead of creating new metrics of each object store type.
> * If you want high-level read/write metrics instead of per operation metrics, you can just do a filter in the grafana dashboard. Example: https://github.com/singularity-data/risingwave/blob/796145df56d0656099609e3df88f5b3d57cd83e8/grafana/risingwave-dashboard.py#L347

Cool. Thanks for the guidelines, I will investigate it later.

File: src/object_store/src/object/mod.rs [hunk 4]
```diff
@@ -140,6 +140,26 @@ pub trait ObjectStore: Send + Sync {
     }
 
     async fn list(&self, prefix: &str) -> ObjectResult<Vec<ObjectMetadata>>;
+
+    fn store_media_type(&self) -> StoreMediaType;
+}
+
+pub enum StoreMediaType {
```
Comment (hzxa21) [inline src/object_store/src/object/mod.rs:147]: This enum looks like a duplicate to ObjectStoreImpl, in which we already differentiate the media type. In addition to using this enum to generate the metric label value, are we planning to use it in other places in the future?

If not, can we just specify a str const in the object store implementation and use it as a label value?

Comment (StrikeW) [reply src/object_store/src/object/mod.rs]: Fixed. seems redundant.

**Review by hzxa21 — APPROVED** [review]
LGTM

**Pull request merged**